### ISSUE TITLE
nn.function.interpolate VERY slow when using AMP

### ISSUE THREAD

**Issue by FabianIsensee** [pytorch/pytorch#51204]
Using nn.functional.interpolate is very slow when using mixed precision training. This seems to have already been an issue in the past (#12409) and should be fixed, but I am still seeing a performance regression of up to 5x (yes, that is 5x the runtime of fp32) depending on the network an GPU.

To demonstrate the issue I have created a small script that reproduces it. I am mostly interested in 3D images which is why that example uses 3D convs, but the same issue appears also with 2D inputs (see below, I will add a separate section for that).

What my script does is the following:
It builds a very simple convolutional encoder network (VGG style) that downsamples the input 1x64x64x64 to 512x4x4x4. It then blows up the small representation back to 512x64x64x64 using interpolate. The loss is then simple the MSE between input and upsampled output. 
This workflow does not represent a useful application, it is just there to put some load on the GPU and isolate the issue.

I run 100 iterations with this setup in three settings
1) regular fp32
2) regular AMP
3) AMP but with a hack: the encoding (512x4x4x4 tensor) is cast to float, then upsampled and then cast to half

Here is the script:

**Comment by FabianIsensee** [comment]
pressed enter too soon, will reopen new issue with the entire texting a couple of minutes
